File: examples/moduledual/native/dual/internal/qwdtt/stats.go
// SPDX-License-Identifier: GPL-3.0-or-later

package qwdtt

import (
	"log"
	"sync/atomic"
	"time"
)

type Stats struct {
	TotalBytesUp      atomic.Int64
	TotalBytesDown    atomic.Int64
	ActiveConnections atomic.Int32
}

func NewStats() *Stats {
	return &Stats{}
}

func (s *Stats) RunLoop(shutdown <-chan struct{}) {
	ticker := time.NewTicker(3 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-shutdown:
			return
		case <-ticker.C:
			active := s.ActiveConnections.Load()
			up := s.TotalBytesUp.Load()
			down := s.TotalBytesDown.Load()
			totalMB := float64(up+down) / (1024.0 * 1024.0)
			upMB := float64(up) / (1024.0 * 1024.0)
			downMB := float64(down) / (1024.0 * 1024.0)

			// Апстрим 1.4.3 добавил разбивку ↓/↑ — берём её, но лейблы остаются АНГЛИЙСКИМИ:
			// это dev-лог (root CLAUDE.md § «Разработческий лог — только английский»).
			log.Printf("[STATS] Active: %d | Traffic: %.2f MB | down %.2f MB / up %.2f MB", active, totalMB, downMB, upMB)
		}
	}
}

File: examples/modulecsqtt/native/csqtt-engine/stats.rs
// SPDX-FileCopyrightText: 2026 amurcanov
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0

use crate::{
    client_perf::{self, Stage as PerfStage},
    events::Events,
};
use std::sync::{
    Arc,
    atomic::{AtomicI32, AtomicI64, Ordering},
};
use tokio_util::sync::CancellationToken;

/// Process-wide mirror of `Stats::active_connections` for the FFI (`csqtt_engine_active_paths`):
/// the helper reads it before every SOCKS dial so it can hold a CONNECT for a few seconds while
/// workers reconnect after sleep instead of burning the 20 s dial timeout.
pub static ACTIVE_PATHS: AtomicI32 = AtomicI32::new(0);

#[derive(Default)]
pub struct Stats {
    pub total_bytes_up: AtomicI64,
    pub total_bytes_down: AtomicI64,
    pub active_connections: AtomicI32,
}

impl Stats {
    pub async fn run(self: Arc<Self>, events: Events, cancel: CancellationToken) {
        let mut interval = tokio::time::interval(std::time::Duration::from_secs(3));
        interval.tick().await;
        let mut last_up = 0i64;
        let mut last_down = 0i64;
        let mut last_active = 0i32;
        let mut quiet_ticks = 0u32;
        loop {
            tokio::select! {
                _ = cancel.cancelled() => return,
                _ = interval.tick() => {
                    client_perf::measure(PerfStage::StatsEmit, || {
                        let active = self.active_connections.load(Ordering::Relaxed);
                        let up = self.total_bytes_up.load(Ordering::Relaxed);
                        let down = self.total_bytes_down.load(Ordering::Relaxed);
                        let unchanged = active == last_active && up == last_up && down == last_down;
                        last_active = active;
                        last_up = up;
                        last_down = down;
                        if unchanged && active == 0 {
                            quiet_ticks = quiet_ticks.saturating_add(1);
                            if quiet_ticks % 10 != 0 {
                                return;
                            }
                        } else {
                            quiet_ticks = 0;
                        }
                        let total_mb = (up + down) as f64 / (1024.0 * 1024.0);
                        crate::log_error!("[СТАТИСТИКА] Активных: {active} | Трафик: {total_mb:.2} МБ");
                        events.stats(active, up, down);
                    });
                }
            }
        }
    }
}
